lib/THC/THCTensorRandom.cu

```diff
@@ -172,15 +177,28 @@ __host__ void THCRandom_getRNGState(THCState* state, THByteTensor *rng_state)
   memcpy(THByteTensor_data(rng_state) + states_size, &state->rngState->current_gen->initial_seed, seed_size);
 }
 
+__global__ void set_rngstate_kernel(curandStateMtgp32 *state, mtgp32_kernel_params *kernel)
+{
+  state[threadIdx.x].k = kernel;
+}
+
 __host__ void THCRandom_setRNGState(THCState* state, THByteTensor *rng_state)
 {
+  if (state->rngState->current_gen == NULL)
+  {
+    THError("Random number generators have not been initialized.");
+  }
+  
   static const size_t states_size = MAX_NUM_BLOCKS * sizeof(curandStateMtgp32);
   static const size_t seed_size = sizeof(unsigned long);
   static const size_t total_size = states_size + seed_size;
   THArgCheck(THByteTensor_nElement(rng_state) == total_size, 1, "RNG state is wrong size");
   THArgCheck(THByteTensor_isContiguous(rng_state), 1, "RNG state must be contiguous");
+  
   THCudaCheck(cudaMemcpy(state->rngState->current_gen->gen_states, THByteTensor_data(rng_state),
                          states_size, cudaMemcpyHostToDevice));
+  set_rngstate_kernel<<<1, MAX_NUM_BLOCKS, 0, THCState_getCurrentStream(state)>>>(
+      state->rngState->current_gen->gen_states, state->rngState->current_gen->kernel_params);

```

Comment:
I think we should synchronize after this call to avoid races between this kernel call and other RNG calls on different streams. We only have one RNG state per device, not per stream. Which, thinking about it, doesn't seem very safe... I couldn't find any information whether of not it's safe in the cuRAND docs though. Does anyone know more about this?